### PR TITLE
Fix two SeekableStream serde issues.

### PR DESCRIPTION
**Pull request by gianm**
1) Fix backwards-compatibility serde for SeekableStreamPartitions. It is needed
   for split 0.13 / 0.14 clusters to work properly during a rolling update.
2) Abstract classes don't need JsonCreator constructors; remove them.

Without this patch, during a rolling update, Kafka index task publishing will not work:

```
2019-03-01T05:37:22,995 WARN [task-runner-0-priority-0] org.apache.druid.indexing.common.actions.RemoteTaskActionClient - Exception submitting action for task[index_kafka_xxx_f8ad14e9e614258_fepajhjj]
org.apache.druid.java.util.common.IOE: Error with status[400 Bad Request] and message[{"error":"Instantiation of [simple type, class org.apache.druid.indexing.kafka.KafkaPartitions] value failed: null (through reference chain: org.apache.druid.indexing.common.actions.TaskActionHolder[\"task\"]->org.apache.druid.indexing.kafka.KafkaIndexTask[\"ioConfig\"]->org.apache.druid.indexing.kafka.KafkaIOConfig[\"startPartitions\"])"}]. Check overlord logs for details.
	at org.apache.druid.indexing.common.actions.RemoteTaskActionClient.submit(RemoteTaskActionClient.java:95) [druid-indexing-service-0.14.0-iap-pre1.jar:0.14.0-iap-pre1]
```

The reason is that the Kafka task object written by an 0.13 MiddleManger is not readable by an 0.14 Overlord. Its `startPartitions` object is perceived as missing required fields by the older overlord.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamPartitions.java [hunk 1]
```diff
@@ -31,14 +31,13 @@
 /**
  * class that encapsulates a partitionIdToSequenceNumberMap of partitionId -> sequenceNumber.
  * To be backward compatible with both Kafka and Kinesis datasource metadata when
- * deserializing json. Redundant constructor fields stream, topic and
+ * serializing and deserializing json, redundant constructor fields stream, topic and
```
Comment (surekhasaharan) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamPartitions.java]: nit: " redundant constructor fields stream, topic, ~and~ partitionSequenceNumberMap and partitionOffsetMap"

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java [hunk 2]
```diff
@@ -44,6 +44,7 @@
                 new NamedType(KafkaIndexTask.class, "index_kafka"),
                 new NamedType(KafkaDataSourceMetadata.class, "kafka"),
                 new NamedType(KafkaIndexTaskIOConfig.class, "kafka"),
+                new NamedType(KafkaIndexTaskTuningConfig.class, "KafkaTuningConfig"),
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: How about `kafka_task_tuning_config`?

[the remark this answers is not in the document]
Comment (surekhasaharan) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: yeah, that sounds good.

[the remark this answers is not in the document]
Comment (gianm) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: Sadly, it needs to be this way for backwards compatibility (the old code didn't have a type name specified, so it used `KafkaTuningConfig`, and we want to be able to read old stuff).

[the remark this answers is not in the document]
Comment (gianm) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: I'll add a comment.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamPartitions.java (hunk 1 above)
Comment (gianm) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamPartitions.java]: Thanks, fixed.

File: extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (surekhasaharan) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: oh, thanks for clarifying and adding comment.

[the remark this answers is not in the document]
Comment (jihoonson) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: Hmm the type name was added in 0.14.0 though (#6431).

[the remark this answers is not in the document]
Comment (jihoonson) [reply extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/indexing/kafka/KafkaIndexTaskModule.java]: Oh I see. Thanks.

**Pull request merged**